runtime/lua/vim/lsp.lua

```diff
@@ -244,6 +244,16 @@ function lsp.start(config, opts)
   local reuse_client = opts.reuse_client or reuse_client_default
   local bufnr = resolve_bufnr(opts.bufnr)
 
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  if
+    vim.bo[bufnr].buftype ~= ''
+    or not vim.api.nvim_buf_is_loaded(bufnr)
+    or not vim.uv.fs_access(fname, 'r')
+  then
+    -- Not a valid buffer
+    return
+  end
```

Comment:
I'm not sure I understand the rationale for this?

Attaching lsp to buffers which are not backed by the filesystem is a legitimate use-case. This would break plugins like nvim-jdtls which support serving `jdt://` uri's via a custom off-spec extension:

https://github.com/mfussenegger/nvim-jdtls/blob/8eb5f0dbe6e126b392ddcaf45893358619893e45/plugin/nvim_jdtls.vim#L6-L7

https://github.com/mfussenegger/nvim-jdtls/blob/8eb5f0dbe6e126b392ddcaf45893358619893e45/lua/jdtls.lua#L1164-L1215

I think https://github.com/sourcegraph/sg.nvim uses a similar approach, and in the LSP repo there are discussions like this: https://github.com/microsoft/language-server-protocol/issues/1264

Reply:
Yes I agree, I often open scratch buffers for Lua stuff that don't get written.

Reply:
Ok this is a case of over fitting to my use case I guess. `clangd` does not work without a backing file so I (incorrectly) assumed that this was the case with all servers.

Will drop the `vim.uv.fs_access` check, but `nvim_buf_is_loaded` and `buftype ~= ''` are still ok I think? Or do either of you know of uses cases where those might be invalid assumptions?

Reply:
`buftype` might be going too far. In my config I just have: `if vim.bo[args.buf].buftype == 'nofile' then`, but that may not be the correct thing to do generally.

Reply:
> Will drop the vim.uv.fs_access check, but nvim_buf_is_loaded and buftype ~= '' are still ok I think? Or do either of you know of uses cases where those might be invalid assumptions?

I second that checking the `buftype` is still too much. Just to show another use-case: https://github.com/mfussenegger/dotfiles/blob/8f99373541a0928eaea7e8e6273307faa370cd34/vim/.config/nvim/commands.lua#L52-L95 there I use `acwrite` to intercept writes to execute code via `BufWriteCmd`. lua-language-server works fine on that scratch buffer.

I think only the `nvim_buf_is_loaded` check is fine, but that's also already present in a way: https://github.com/neovim/neovim/blob/a1550dbf0ac252052a7a71780708d3ec05aa3686/runtime/lua/vim/lsp.lua#L596-L599

Reply:
What about disallowing `nofile`? I can't remember why I needed it, but it was to prevent attaching to some intermediate buffers, probably gitsigns diff buffers.

Reply:
`nofile` definitely works. I use that in nvim-jdtls.

I think the only buftype that might cause problems is `prompt`, but that might also have been fixed with https://github.com/neovim/neovim/pull/28260